### PR TITLE
fix: fix ControlPlane managed resource labels during migration from 1.2 and older

### LINKED ISSUE #373: ControlPlane cluster resource do not allow migration from <1.2 (owner ref link) to 1.3+ (label based owner ref)
### Problem statement

#259 introduced a new way to link `ControlPlane`'s cluster wide resources to the managing `ControlPlane` via labels.

This approach works but it misses the migration approach:

- users install `ControlPlane`s using KGO 1.2 (or older)
- KGO creates owner reference bindings between `ControlPlane`'s cluster wide resources to the managing `ControlPlane`
- users upgrade KGO to 1.3
- KGO looks up the cluster wide resources via the new, label based approach, and misses the old ones that are already created, and thus tries to create the already existing resources

Logs:

```
manager {"level":"debug","ts":"2024-06-24T09:47:57Z","logger":"controlplane","msg":"admission webhook enabled, enforcing admission webhook resources","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","ControlPlane":{"name":"kong-2r5z8","namespace":"default"},"namespace":"default","name":"kong-2r5z8","reconcileID":"44661c34-2d58-418f-afbf-cf2807687d6e","namespace":"default","name":"kong-2r5z8"}
manager {"level":"Level(-2)","ts":"2024-06-24T09:47:57Z","logger":"controlplane","msg":"ensuring admission webhook service","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","ControlPlane":{"name":"kong-2r5z8","namespace":"default"},"namespace":"default","name":"kong-2r5z8","reconcileID":"44661c34-2d58-418f-afbf-cf2807687d6e","namespace":"default","name":"kong-2r5z8"}
manager {"level":"Level(-2)","ts":"2024-06-24T09:47:57Z","logger":"controlplane","msg":"ensuring admission webhook certificate","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","ControlPlane":{"name":"kong-2r5z8","namespace":"default"},"namespace":"default","name":"kong-2r5z8","reconcileID":"44661c34-2d58-418f-afbf-cf2807687d6e","namespace":"default","name":"kong-2r5z8"}
manager {"level":"Level(-2)","ts":"2024-06-24T09:47:57Z","logger":"controlplane","msg":"ensuring admission webhook configuration","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","ControlPlane":{"name":"kong-2r5z8","namespace":"default"},"namespace":"default","name":"kong-2r5z8","reconcileID":"44661c34-2d58-418f-afbf-cf2807687d6e","namespace":"default","name":"kong-2r5z8"}
manager {"level":"error","ts":"2024-06-24T09:47:57Z","msg":"Reconciler error","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","ControlPlane":{"name":"kong-2r5z8","namespace":"default"},"namespace":"default","name":"kong-2r5z8","reconcileID":"44661c34-2d58-418f-afbf-cf2807687d6e","error":"failed to ensure webhook resources: validatingwebhookconfigurations.admissionregistration.k8s.io \"kong-2r5z8\" already exists","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222"}
```

This causes e2e upgrade tests to fail.

### Steps to reproduce:

- Install KGO 1.2 using a helm chart
- Apply Gateway manifest e.g. https://github.com/Kong/gateway-operator/blob/fa93ae9b150388d18fd4d6e01bcceb77736aa57a/config/samples/gateway-with-gatewayconfiguration.yaml
- Update KGO to 1.3 (or nightly at this point)
- Observe the above mentioned logs and that `Gateway` does not get `Programmed`

### Anything else ?

Old labels

```
	gateway-operator.konghq.com/managed-by: controlplane
	konghq.com/gateway-operator: controlplane
```

New labels

```
	gateway-operator.konghq.com/managed-by: controlplane
	gateway-operator.konghq.com/managed-by-name: kong-ncv7m
	gateway-operator.konghq.com/managed-by-namespace: default
	konghq.com/gateway-operator: controlplane
```

Since we're using the new label set for both the generation and existing resources lookup, existing objects (created with old labels set) are not matched and hence a create operation is performed which should be an update to migrate to the new set of labels.

### Acceptance criteria

- [ ] When KGO is upgraded from 1.2 to 1.3 `ControlPlane` managed cluster wide resources have their labels properly migrated (updated) and are properly looked up
- [ ] E2E upgrade test assertion is added to https://github.com/Kong/gateway-operator/blob/0781df268b2b75a70895c46c6bf5eed5e08f8887/test/e2e/test_helm_install_upgrade.go and ensures that the above happens.

### PR DESCRIPTION
**Pull request by pmalek**
**What this PR does / why we need it**:

#259 introduced a new set of labels that are used on `ControlPlane`'s
- `ClusterRole`s
- `ClusterRoleBindings`s
- `ValidatingWebhookConfiguration`s

This omitted 1 important detail which is migrating on operator upgrade (specifically from 1.2.x and older to 1.3.0).

This PR fixes this by changing the listing and deleting logic for the aforementioned resources so that the old label and owner reference are also taken into account.

This also makes other small changes
- make operator leave the newest `ClusterRole` or `ValidatingWebhookConfiguration` in place when there's more in the cluster;
  - this ensures that when for some reason there are more than 1 of these, the newest with newest rules, stays
- when `ControlPlane` fails to get provisioned - this can happen when operator's webhook doesn't respond - requeue, without this, provisioned `Gateway` gets stuck with `ControlPlaneReady` condition set to `False`.

**Which issue this PR fixes**

Fixes #373 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: controller/controlplane/controller.go [hunk 1]
```diff
@@ -63,9 +63,6 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 	validatinWebhookConfigurationOwnerPredicate.UpdateFunc = func(e event.UpdateEvent) bool {
 		return r.validatingWebhookConfigurationHasControlPlaneOwner(e.ObjectOld)
 	}
-	validatinWebhookConfigurationOwnerPredicate.DeleteFunc = func(e event.DeleteEvent) bool {
-		return r.validatingWebhookConfigurationHasControlPlaneOwner(e.Object)
-	}
```
Comment (pmalek) [inline controller/controlplane/controller.go:68]: This is already set by calling `predicate.NewPredicateFuncs`

File: controller/controlplane/controlplane_controller_reconciler_utils_test.go [hunk 2]
```diff
@@ -82,12 +81,12 @@ func TestEnsureClusterRole(t *testing.T) {
...
-			controlplane:        controlplane,
-			createdorUpdated:    true,
-			expectedClusterRole: *clusterRole,
-		},
+		// {
+		// 	Name:                "no existing clusterrole",
+		// 	controlplane:        controlplane,
+		// 	createdorUpdated:    true,
+		// 	expectedClusterRole: *clusterRole,
+		// },
```
Comment (programmer04) [inline controller/controlplane/controlplane_controller_reconciler_utils_test.go]: ???

File: controller/controlplane/controller_watch.go [hunk 3]
```diff
@@ -146,23 +146,48 @@ func (r *Reconciler) getControlPlaneRequestFromManagedByNameLabel(ctx context.Co
...
+
+	if labels := obj.GetLabels(); len(labels) > 0 {
+		if labels[consts.GatewayOperatorManagedByNameLabel] == cp.Name {
+			if obj.GetNamespace() != "" {
+				return cp.GetNamespace() == obj.GetNamespace()
+			} else {
+				return true
+			}
+		}
+	}
```
Comment (programmer04) [inline controller/controlplane/controller_watch.go]:
```suggestion
	if labels[consts.GatewayOperatorManagedByNameLabel] == cp.Name {
		if obj.GetNamespace() != "" {
			return cp.GetNamespace() == obj.GetNamespace()
		} else {
			return true
		}
	}
```

Reading from empty/nil map is safe it returns default object value in this case empty string - `""`. The same as for non-existing item. So the first check can be dropped without changing semantic of this snippet

File: pkg/utils/kubernetes/reduce/filters.go [hunk 4]
```diff
@@ -86,20 +86,49 @@ func filterServiceAccounts(serviceAccounts []corev1.ServiceAccount) []corev1.Ser
...
+			newestLegacy = i
+		}
+		continue
 	}
 
-	return append(clusterRoles[:toFilter], clusterRoles[toFilter+1:]...)
+	if newestWithManagedByLabels != -1 {
+		return append(clusterRoles[:newestWithManagedByLabels], clusterRoles[newestWithManagedByLabels+1:]...)
+	}
+	return append(clusterRoles[:newestLegacy], clusterRoles[newestLegacy+1:]...)
```
Comment (programmer04) [inline pkg/utils/kubernetes/reduce/filters.go:131]: Here we're exposed to issue described in 
- https://github.com/Kong/gateway-operator/issues/146

File: pkg/utils/kubernetes/reduce/filters.go [hunk 5]
```diff
@@ -115,14 +144,41 @@ func filterClusterRoleBindings(clusterRoleBindings []rbacv1.ClusterRoleBinding)
...
+			oldestLegacy = i
+		}
+		continue
 	}
 
-	return append(clusterRoleBindings[:toFilter], clusterRoleBindings[toFilter+1:]...)
+	if oldestWithManagedByLabels != -1 {
+		return append(clusterRoleBindings[:oldestWithManagedByLabels], clusterRoleBindings[oldestWithManagedByLabels+1:]...)
+	}
+	return append(clusterRoleBindings[:oldestLegacy], clusterRoleBindings[oldestLegacy+1:]...)
```
Comment (programmer04) [inline pkg/utils/kubernetes/reduce/filters.go:181]: Here we're exposed to the issue described in 
- https://github.com/Kong/gateway-operator/issues/146

File: pkg/utils/kubernetes/reduce/filters.go [hunk 6]
```diff
@@ -315,21 +371,52 @@ func FilterHPAs(hpas []autoscalingv2.HorizontalPodAutoscaler) []autoscalingv2.Ho
...
+			newestLegacy = i
+		}
+		continue
 	}
 
-	return append(webhookConfigurations[:best], webhookConfigurations[best+1:]...)
+	if newestWithManagedByLabels != -1 {
+		return append(vwcs[:newestWithManagedByLabels], vwcs[newestWithManagedByLabels+1:]...)
+	}
+	return append(vwcs[:newestLegacy], vwcs[newestLegacy+1:]...)
```
Comment (programmer04) [inline pkg/utils/kubernetes/reduce/filters.go:419]: Here we're exposed to issue described in 
- https://github.com/Kong/gateway-operator/issues/146

File: controller/controlplane/controlplane_controller_reconciler_utils_test.go (hunk 2 above)
Comment (pmalek) [reply controller/controlplane/controlplane_controller_reconciler_utils_test.go]: oops, 30a8769ab05b6b6b287a10aed2ebe1344c32788a

File: pkg/utils/kubernetes/reduce/filters.go (hunk 4 above)
Comment (pmalek) [reply pkg/utils/kubernetes/reduce/filters.go:131]: ACK, let's keep the scope of this PR as small as possible. 146 deserves its own issue. WDYT?

**Review by programmer04 — APPROVED** [review]
LGTM, but since it will be dealing with existing users configs I think someone else should take a look too, to double check

File: controller/controlplane/controller_reconciler_utils.go [hunk 7]
```diff
@@ -293,16 +293,41 @@ func (r *Reconciler) ensureClusterRole(
 	ctx context.Context,
 	cp *operatorv1beta1.ControlPlane,
 ) (createdOrUpdated bool, cr *rbacv1.ClusterRole, err error) {
-	managedByLabelSet := k8sutils.GetManagedByLabelSet(cp)
+	// TODO: This performs a migration from the old managedBy label to the new one.
+	// After several versions of soak time we can remove handling the legeacy label set.
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// After several versions of soak time we can remove handling the legacy label set.
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 8]
```diff
@@ -293,16 +293,41 @@ func (r *Reconciler) ensureClusterRole(
...
+	// PR that introduced the new set of labels: https://github.com/Kong/gateway-operator/pull/259
+	// PR: that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// PR that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 9]
```diff
@@ -350,17 +375,41 @@ func (r *Reconciler) ensureClusterRoleBinding(
 ) (createdOrUpdate bool, crb *rbacv1.ClusterRoleBinding, err error) {
 	logger := log.GetLogger(ctx, "controlplane.ensureClusterRoleBinding", r.DevelopmentMode)
 
-	managedByLabelSet := k8sutils.GetManagedByLabelSet(cp)
+	// TODO: This performs a migration from the old managedBy label to the new one.
+	// After several versions of soak time we can remove handling the legeacy label set.
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// After several versions of soak time we can remove handling the legacy label set.
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 10]
```diff
@@ -350,17 +375,41 @@ func (r *Reconciler) ensureClusterRoleBinding(
...
+	// PR that introduced the new set of labels: https://github.com/Kong/gateway-operator/pull/259
+	// PR: that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// PR that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 11]
```diff
@@ -671,15 +766,42 @@ func (r *Reconciler) ensureValidatingWebhookConfiguration(
 ) (op.Result, error) {
 	logger := log.GetLogger(ctx, "controlplane.ensureValidatingWebhookConfiguration", r.DevelopmentMode)
 
-	managedByLabelSet := k8sutils.GetManagedByLabelSet(cp)
+	// TODO: This performs a migration from the old managedBy label to the new one.
+	// After several versions of soak time we can remove handling the legeacy label set.
+	// PR that introduced the new set of labels: https://github.com/Kong/gateway-operator/pull/259
+	// PR: that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// PR that introduced the migration: https://github.com/Kong/gateway-operator/pull/369
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 12]
```diff
@@ -671,15 +766,42 @@ func (r *Reconciler) ensureValidatingWebhookConfiguration(
 ) (op.Result, error) {
 	logger := log.GetLogger(ctx, "controlplane.ensureValidatingWebhookConfiguration", r.DevelopmentMode)
 
-	managedByLabelSet := k8sutils.GetManagedByLabelSet(cp)
+	// TODO: This performs a migration from the old managedBy label to the new one.
+	// After several versions of soak time we can remove handling the legeacy label set.
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]:
```suggestion
	// After several versions of soak time we can remove handling the legacy label set.
```

File: controller/controlplane/controller_reconciler_utils.go [hunk 13]
```diff
@@ -671,15 +766,42 @@ func (r *Reconciler) ensureValidatingWebhookConfiguration(
 ) (op.Result, error) {
 	logger := log.GetLogger(ctx, "controlplane.ensureValidatingWebhookConfiguration", r.DevelopmentMode)
 
-	managedByLabelSet := k8sutils.GetManagedByLabelSet(cp)
+	// TODO: This performs a migration from the old managedBy label to the new one.
```
Comment (mlavacca) [inline controller/controlplane/controller_reconciler_utils.go]: Is the migration performed here? Is the new label set enforced or the migration is something we still need to address?

File: controller/controlplane/controller_watch.go [hunk 14]
```diff
@@ -74,18 +75,17 @@ func (r *Reconciler) validatingWebhookConfigurationHasControlPlaneOwner(obj clie
 // ClusterScopedObjHasControlPlaneOwner checks if the cluster-scoped object has a control plane owner.
 // The check is performed through the managed-by-name label.
 func (r *Reconciler) ClusterScopedObjHasControlPlaneOwner(ctx context.Context, obj client.Object) bool {
-	controlplaneList := &operatorv1beta1.ControlPlaneList{}
-	if err := r.Client.List(ctx, controlplaneList); err != nil {
+	var controlplaneList operatorv1beta1.ControlPlaneList
```
Comment (mlavacca) [inline controller/controlplane/controller_watch.go:78]: What's the point of this change?

File: controller/controlplane/controller_watch.go [hunk 15]
```diff
@@ -146,23 +146,47 @@ func (r *Reconciler) getControlPlaneRequestFromManagedByNameLabel(ctx context.Co
...
+			},
 		}
 	}
 
 	return
 }
 
+// objectIsOwnedByControlPlane checks if the object is owned by the control plane.
+//
+// NOTE: We are using the managed-by-name label to identify the owner of the resource
```
Comment (mlavacca) [inline controller/controlplane/controller_watch.go]:
```suggestion
// NOTE: We are using the managed-by-name label to identify the owner of the resource.
```

File: controller/controlplane/controller_watch.go [hunk 16]
```diff
@@ -146,23 +146,47 @@ func (r *Reconciler) getControlPlaneRequestFromManagedByNameLabel(ctx context.Co
...
+// To keep backward compatibility, we also check the owner reference which
+// are not used anymore for cluster-scoped resources since that's considered
```
Comment (mlavacca) [inline controller/controlplane/controller_watch.go]:
```suggestion
// is not used anymore for cluster-scoped resources since that's considered
```

File: controller/gateway/controller.go [hunk 17]
```diff
@@ -309,11 +309,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-		condition, found := k8sutils.GetCondition(ControlPlaneReadyType, oldGwConditionsAware)
-		if !found || condition.Status == metav1.ConditionTrue {
+	if condition, found := k8sutils.GetCondition(ControlPlaneReadyType, gwConditionAware); found && condition.Status != metav1.ConditionTrue {
+		if condition.Reason == string(consts.UnableToProvisionReason) {
+			log.Debug(logger, "unable to provision controlplane, requeueing", gateway)
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		conditionOld, foundOld := k8sutils.GetCondition(ControlPlaneReadyType, oldGwConditionsAware)
+		if !foundOld || conditionOld.Status == metav1.ConditionTrue {
```
Comment (mlavacca) [inline controller/gateway/controller.go:321]: I'm pretty scared by these changes as the Gateway status historically has been a delicate point to touch. What's the root cause of this change here? Any bug we need to fix?

File: controller/controlplane/controller_watch.go (hunk 14 above)
Comment (pmalek) [reply controller/controlplane/controller_watch.go:78]: No need to allocate here. We can use a var on stack. (note: this is probably going to be optimized in the compiler during the escape analysis: https://tip.golang.org/src/cmd/compile/internal/escape/escape.go, but if we can let's mark this as stack allocated anyway).

WDYT?

File: controller/controlplane/controller_reconciler_utils.go (hunk 13 above)
Comment (pmalek) [reply controller/controlplane/controller_reconciler_utils.go]: This lists the old and the new (below) `ValidatingWebhookConfiguration`s, and then reduces those to just 1 (if needed) and updates the object in cluster (if needed below).

The TODO in here is to remember to remove the "migration" at some point. 

I've created #401 to track this.

File: controller/gateway/controller.go (hunk 17 above)
Comment (pmalek) [reply controller/gateway/controller.go:321]: The only change in here is in the lines 315-318 which requeues when the CP failed to get provisioned. This could happen (as described in the PR description) when e.g. webhook is not available (this can happen when operator is upgraded).

Without this requeue the reconciliation gets stuck and the only indication of such a situation is a condition on the Gateway.

File: controller/controlplane/controller_watch.go (hunk 14 above)
Comment (mlavacca) [reply controller/controlplane/controller_watch.go:78]: ok, it makes sense 👍

File: controller/controlplane/controller_reconciler_utils.go (hunk 13 above)
Comment (mlavacca) [reply controller/controlplane/controller_reconciler_utils.go]: I see. Could you rephrase the comment to clearly state this aspect? It's not immediate to me that the TODO is related to removing the migration logic. Also, mentioning the #401 in the code would be great. WDYT?

Comment (pmalek) [reply controller/controlplane/controller_reconciler_utils.go]: 423c792f64b0fd4ff9a6c273fe2a42614a9aacca PTAL

Comment (mlavacca) [reply controller/controlplane/controller_reconciler_utils.go]: that's great 👍

**Review by mlavacca — APPROVED** [review]
